File: app/src/main/java/com/android/ashwiask/tvmaze/shows/ShowsDataSource.java
package com.android.ashwiask.tvmaze.shows;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.paging.ItemKeyedDataSource;
import android.support.annotation.NonNull;
import android.util.Log;

import com.android.ashwiask.tvmaze.home.Show;
import com.android.ashwiask.tvmaze.home.TvMazeApi;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

@Singleton
public class ShowsDataSource extends ItemKeyedDataSource<Integer, Show> {
    private static final String TAG = "ShowsDataSource";
    private final TvMazeApi tvMazeApi;
    private CompositeDisposable compositeDisposable;
    private int pageNumber = 1;
    private MutableLiveData<NetworkState> paginatedNetworkStateLiveData;
    private MutableLiveData<NetworkState> initialLoadStateLiveData;

    @Inject
    public ShowsDataSource(TvMazeApi tvMazeApi) {
        this.tvMazeApi = tvMazeApi;
        compositeDisposable = new CompositeDisposable();
        initialLoadStateLiveData = new MutableLiveData<>();
        paginatedNetworkStateLiveData = new MutableLiveData<>();
    }

    @Override
    public void loadInitial(@NonNull LoadInitialParams<Integer> params,
                            @NonNull LoadInitialCallback<Show> callback) {
        Log.d(TAG, "Fetching first page: " + pageNumber);
        initialLoadStateLiveData.postValue(NetworkState.builder()
                .status(NetworkState.Status.LOADING).build());
        Disposable showsDisposable = tvMazeApi.getShows(pageNumber)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shows -> onShowsFetched(shows, callback), this::onError);
        compositeDisposable.add(showsDisposable);
    }

    private void onError(Throwable throwable) {
        initialLoadStateLiveData.postValue(NetworkState.builder()
                .status(NetworkState.Status.ERROR)
                .message(throwable.getMessage()).build());
        Log.d(TAG, throwable.getMessage());
    }

    private void onShowsFetched(List<Show> shows, LoadInitialCallback<Show> callback) {
        initialLoadStateLiveData.postValue(NetworkState.builder()
                .status(NetworkState.Status.SUCCESS).build());
        pageNumber++;
        callback.onResult(shows);
    }

    @Override
    public void loadAfter(@NonNull LoadParams<Integer> params,
                          @NonNull LoadCallback<Show> callback) {
        Log.d(TAG, "Fetching next page: " + pageNumber);
        paginatedNetworkStateLiveData.postValue(NetworkState.builder()
                .status(NetworkState.Status.LOADING).build());
        Disposable showsDisposable = tvMazeApi.getShows(params.key)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(shows -> onMoreShowsFetched(shows, callback), this::onPaginationError);
        compositeDisposable.add(showsDisposable);
    }

    private void onMoreShowsFetched(List<Show> shows, LoadCallback<Show> callback) {
        paginatedNetworkStateLiveData.postValue(NetworkState.builder()
                .status(NetworkState.Status.SUCCESS).build());
        pageNumber++;
        callback.onResult(shows);
    }

    private void onPaginationError(Throwable throwable) {
        paginatedNetworkStateLiveData.postValue(NetworkState.builder()
                .status(NetworkState.Status.ERROR)
                .message(throwable.getMessage()).build());
        Log.d("ShowsDataSource", throwable.getMessage());
    }

    @Override
    public void loadBefore(@NonNull LoadParams<Integer> params,
                           @NonNull LoadCallback<Show> callback) {
        // Do nothing, since data is loaded from our initial load itself
    }

    @NonNull
    @Override
    public Integer getKey(@NonNull Show item) {
        return pageNumber;
    }

    public void clear() {
        compositeDisposable.clear();
        pageNumber = 1;
    }

    public LiveData<NetworkState> getPaginatedNetworkStateLiveData() {
        return paginatedNetworkStateLiveData;
    }

    public LiveData<NetworkState> getInitialLoadStateLiveData() {
        return initialLoadStateLiveData;
    }
}

File: app/src/main/java/com/android/ashwiask/tvmaze/shows/ShowsDataSource.kt
package com.android.ashwiask.tvmaze.shows

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.paging.ItemKeyedDataSource
import com.android.ashwiask.tvmaze.favorite.FavoriteShowsRepository
import com.android.ashwiask.tvmaze.home.TvMazeApi
import com.android.ashwiask.tvmaze.network.home.Show
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ShowsDataSource @Inject
constructor(
    private val tvMazeApi: TvMazeApi,
    private val favoriteShowsRepository: FavoriteShowsRepository
) : ItemKeyedDataSource<Int, Show>() {
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()
    private var pageNumber = 1
    private val paginatedNetworkStateLiveData: MutableLiveData<NetworkState> = MutableLiveData()
    private val initialLoadStateLiveData: MutableLiveData<NetworkState> = MutableLiveData()
    // For Retry
    private var params: LoadParams<Int>? = null
    private var callback: LoadCallback<Show>? = null
    private val favoriteShows: Single<List<Show>>? = null

    override fun loadInitial(
        params: LoadInitialParams<Int>,
        callback: LoadInitialCallback<Show>
    ) {
        Log.d(TAG, "Fetching first page: $pageNumber")
        initialLoadStateLiveData.postValue(Loading)
        val showsDisposable = tvMazeApi.getShows(pageNumber)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({ onShowsFetched(it, callback) }, { onError(it) })
        compositeDisposable.add(showsDisposable)
    }

    private fun onError(throwable: Throwable) {
        initialLoadStateLiveData.postValue(NetworkError(throwable.message))
    }

    private fun onShowsFetched(
        shows: List<Show>,
        callback: LoadInitialCallback<Show>
    ) {
        initialLoadStateLiveData.postValue(Success)
        pageNumber++
        callback.onResult(shows)
    }

    override fun loadAfter(
        params: LoadParams<Int>,
        callback: LoadCallback<Show>
    ) {
        this.params = params
        this.callback = callback
        Log.d(TAG, "Fetching next page: $pageNumber")
        paginatedNetworkStateLiveData.postValue(Loading)
        val showsDisposable = tvMazeApi.getShows(params.key)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                { onMoreShowsFetched(it, callback) },
                { this.onPaginationError(it) })
        compositeDisposable.add(showsDisposable)
    }

    private fun onMoreShowsFetched(shows: List<Show>, callback: LoadCallback<Show>) {
        paginatedNetworkStateLiveData.postValue(Success)
        pageNumber++
        callback.onResult(shows)
    }

    private fun onPaginationError(throwable: Throwable) {
        paginatedNetworkStateLiveData.postValue(NetworkError(throwable.message))
        Log.e(TAG, throwable.message)
    }

    override fun loadBefore(
        params: LoadParams<Int>,
        callback: LoadCallback<Show>
    ) {
        // Do nothing, since data is loaded from our initial load itself
    }

    override fun getKey(item: Show): Int {
        return pageNumber
    }

    fun clear() {
        compositeDisposable.clear()
        pageNumber = 1
    }

    fun getPaginatedNetworkStateLiveData(): LiveData<NetworkState> {
        return paginatedNetworkStateLiveData
    }

    fun getInitialLoadStateLiveData(): LiveData<NetworkState> {
        return initialLoadStateLiveData
    }

    fun retryPagination() {
        loadAfter(params!!, callback!!)
    }

    companion object {
        private const val TAG = "ShowsDataSource"
    }
}
